File: training-data/go/kmeans_clustering.go
package main

import (
	"fmt"
	"math"
)

type point struct {
	x, y float64
}

func distance(a, b point) float64 {
	return math.Sqrt((a.x-b.x)*(a.x-b.x) + (a.y-b.y)*(a.y-b.y))
}

func kmeans(points []point, k, iterations int) []point {
	centroids := append([]point(nil), points[:k]...)

	for iter := 0; iter < iterations; iter++ {
		clusters := make([][]point, k)
		for _, p := range points {
			closest := 0
			minDist := math.MaxFloat64
			for i := 0; i < k; i++ {
				d := distance(p, centroids[i])
				if d < minDist {
					minDist = d
					closest = i
				}
			}
			clusters[closest] = append(clusters[closest], p)
		}

		for i, cluster := range clusters {
			if len(cluster) == 0 {
				continue
			}
			var sumX, sumY float64
			for _, p := range cluster {
				sumX += p.x
				sumY += p.y
			}
			centroids[i] = point{sumX / float64(len(cluster)), sumY / float64(len(cluster))}
		}
	}
	return centroids
}

func main() {
	points := []point{{1, 1}, {1.5, 2}, {3, 4}, {5, 7}, {3.5, 5}, {4.5, 5}, {3.5, 4.5}}
	fmt.Println(kmeans(points, 2, 20))
}

File: training-data/python/kmeans_clustering.py
import random
from typing import List, Tuple

Point = Tuple[float, float]


def distance(a: Point, b: Point) -> float:
    return ((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2) ** 0.5


def kmeans(points: List[Point], k: int, iterations: int = 20) -> List[Point]:
    centroids = random.sample(points, k)

    for _ in range(iterations):
        clusters: List[List[Point]] = [[] for _ in range(k)]
        for point in points:
            closest = min(range(k), key=lambda i: distance(point, centroids[i]))
            clusters[closest].append(point)

        new_centroids = []
        for i, cluster in enumerate(clusters):
            if cluster:
                mean_x = sum(p[0] for p in cluster) / len(cluster)
                mean_y = sum(p[1] for p in cluster) / len(cluster)
                new_centroids.append((mean_x, mean_y))
            else:
                new_centroids.append(centroids[i])
        centroids = new_centroids

    return centroids


if __name__ == "__main__":
    random.seed(42)
    points = [(1, 1), (1.5, 2), (3, 4), (5, 7), (3.5, 5), (4.5, 5), (3.5, 4.5)]
    print(kmeans(points, 2))
